File: 22. Generate Parentheses/solution.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        response = []
        def backtrack(s=[],left =0 , right = 0):
            if len(s) == 2 * n:
                response.append("".join(s))
                return
            if left < n:
                s.append("(")
                backtrack(s,left+1,right)
                s.pop()
            if right < left:
                s.append(")")
                backtrack(s,left, right + 1)
                s.pop()
        backtrack()
        return response

File: 22. Generate Parentheses/solution.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    let solution = []
    const max_p = n * 2
    const backtracking = (actual=[],left=0,right=0) =>{
        if (actual.length ===  max_p){
            solution.push(actual.join(""))
        }
        if(left < n){
            actual.push("(")
            backtracking(actual,left+1,right)
            actual.pop()
        }
        if(right < left){
            actual.push(")")
            backtracking(actual,left,right+1)
            actual.pop()
        }
    }
    backtracking()
    return solution
};
